### ISSUE TITLE
Rework t.w.p.jabber.component to work with initializers.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1995]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#1995|
|Type|enhancement|
|Created|2006-08-13 12:14:31Z|

After work on [#10701](https://github.com/twisted/twisted/issues/10701) has been finished, the handshaking for server components can also be converted to use the new initializer concept.

Attachments:

* [handshake_as_initilizer.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/87e/87e8db4f2338ba69baa1c7d4e60969caf4f06d9e/a98c18705726b426d67ca82f459d5c5e7e83434e.patch) (3168 bytes) - added by ralphm on 2006-08-13 12:15:57Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1995 1995
type__enhancement enhancement
reporter__ralphm ralphm
priority__highest highest
milestone__Words_0_5 Words-0.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1155471271000000 1155471271000000
changetime__1161270841000000 1161270841000000
version__None None
owner__ 
cc__ralphm cc__glyph cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_1"></a>|@ralphm removed owner|
|-|-|

Applied patch and added doctests in [18388]. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm commented|
|-|-|

For completeness: [branches/jabber-component-initializers-1995](branches/jabber-component-initializers-1995)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @ralphm<br>@glyph set status to `new`|
|-|-|

It says "added doctests" in your comment, but I don't see doctests, I see docstrings.  That's good, though, Twisted's testing standard is _not_ doctests, it's unit tests.  Unfortunately I don't see any of those, either.  Please write a few to cover the new functionality and re-review.

It might be worth mentioning here that it's strongly suggested to develop new code for Twisted test-first, if you can.  In many situations that's really hard, but it seems like you know the domain pretty well here, so in the future try to write the tests before the implementation (it's a lot easier to see the coverage).

The module docstring should really contain at least a brief explanation (and a pointer to a longer one, if there is one) of what the word "external" and "component" mean.  This is especially important because other areas of Twisted use these same words to mean other things: "external" meaning "not within the Twisted codebase" and "component" to mean "composed object implementing a Zope-style interface".

Some parameter names are changed here from "xmlstream" to "xs".  This seems less clear, and is especially problematic since there is no ```@param``` in the appropriate docstrings explaining what the parameter was for.  Call the parameter whatever you like as long as you have some documentation for it :).

The new docstring for "send" doesn't describe what "obj" is, and in my experience parameters with names like "obj" or "thunk" or "thing" are the ones that most desperately need clear documentation.  What kind of object is "obj"?

Not a comment on this branch in particular, but there are a few problems with ```IService```:

 * Its docstring doesn't explain its role or intended usage.
 * Its name conflicts with ```twisted.application.service.IService```, which is an extremely common interface within Twisted.
 * It has only one implementor so far.  There should be at least an example of a second implementation if it is an interface that users are supposed to use.

There's also a convention in the Divmod repository which we may want to adopt, to put all new interfaces into special "interfaces only" modules, so they can be imported without sucking in the whole implementation.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_5"></a>|@ralphm commented|
|-|-|

Sure, I meant docstrings. There is no new functionality, I just redid the handshake stuff such that it uses the new initializer stuff. The existing test `ComponentAuthTest` in `test_jabbercomponent.py` covers it.

I'll add some text on external components.

I changed `xmlstream` to `xs`, because `xmlstream` overrides an import and I use `xs` througout recently touched code. I'll add docstrings to compensate.

On `send`, I'll to the docstrings.

On `IService`, the interface is there because ServerManager invokes methods on its child services that implement this interface. I'll clarify this in the docstring. About the name, do you have a suggestion for a better name? Also, is component.Service, the only implementation of component.IService also a less fortunate name?

I would be ok with creating a interfaces.py, would it also go in t.w.p.jabber?

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_6"></a>|@ralphm removed owner|
|-|-|

Improved docstrings.

glyph: in the future I might even spin out the `Service` and `IService` for use with similar ServiceManagers for server-to-server or client-to-server connections.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @ralphm<br>@exarkun set status to `new`|
|-|-|

The added docstrings are excellent.

The code being refactored is mainly covered by existing tests, although some of it only indirectly (eg, there are no unit tests for ```ComponentInitiatingInitializer```, though all of its lines are at least exercised indirectly by other tests).  There doesn't seem to have been any existing test coverage for ```Service```.  However, since ```Service``` isn't new in this branch, I don't know that this lack needs to block the merge of this branch (if it doesn't, though, please create a ticket for adding test coverage).

I think ```IComponentService``` would be an improvement on ```IService```.  Likewise, if there is a default implementation, ```ComponentService``` makes more sense to me.  This is generally keeping with the naming convention used through the rest of Twisted (ie, ```twisted.mail.imap4.IMAP4Server```).

A convention which seems to be emerging is for interface modules to be named ```ipackage```.  For example, ```twisted.words.jabber.ijabber```.  This hasn't really solidifed yet, and it isn't used as widely through Twisted yet (though newer interface modules seem to be following it at least some of the time).  Either ```ijabber``` or ```interfaces``` would be fine with me.

Given the amount of code in ```Service```, perhaps it doesn't even need to exist.  Given that it wasn't added in this branch though, it sounds like another change to be made in another branch.  Removing it should put glyph's concern that there is only one implementation to rest, too.

So in summary, add unit tests for ```ComponentInitiatingInitializer```, create some tickets for the other fixes, and then merge.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_9"></a>|@ralphm commented|
|-|-|

 * Regarding the location of `Interface` definitions, I created [#2177](https://github.com/twisted/twisted/issues/2177).
 * Regarding the name and purpose of `IService`, I created [#2178](https://github.com/twisted/twisted/issues/2178).
 * Added tests for `ComponentInitiatingInitializer`.

Merging.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_10"></a>|@ralphm set status to `closed`|
|-|-|

(In [18481]) Refactor component handshake as stream initializer.

Author: ralphm
Reviewer: exarkun, glyph
Fixes [#1995](https://github.com/twisted/twisted/issues/1995)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
